NPE when casting map containing NULLs

```sql
SELECT CAST(MAP(ARRAY[0, 1], ARRAY[1,NULL]) AS MAP<BIGINT, DOUBLE>);
```

```
com.facebook.presto.spi.PrestoException: GENERIC_INTERNAL_ERROR
	at com.facebook.presto.operator.scalar.MapToMapCast.toMap(MapToMapCast.java:110)
	at com_facebook_presto_$gen_PageProcessor_27357.project_0(Unknown Source)
	at com_facebook_presto_$gen_PageProcessor_27357.process(Unknown Source)
	at com.facebook.presto.operator.FilterAndProjectOperator.getOutput(FilterAndProjectOperator.java:111)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:378)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:301)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:622)
	at com.facebook.presto.execution.TaskExecutor$PrioritizedSplitRunner.process(TaskExecutor.java:529)
	at com.facebook.presto.execution.TaskExecutor$Runner.run(TaskExecutor.java:665)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
	at sun.invoke.util.ValueConversions.unboxLong(ValueConversions.java:121)
	at com.facebook.presto.operator.scalar.MapToMapCast.toMap(MapToMapCast.java:104)
	... 11 more
```

--- a/presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java
+++ b/presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java
@@ -99,6 +99,11 @@ public final class MapToMapCast
             if (!typedSet.contains(keyBlock, i / 2)) {
                 typedSet.add(keyBlock, i / 2);
                 toKeyType.appendTo(keyBlock, i / 2, blockBuilder);
+                if (fromMap.isNull(i + 1)) {
+                    blockBuilder.appendNull();
+                    continue;
+                }
+
                 Object fromValue = readNativeValue(fromValueType, fromMap, i + 1);
                 try {
                     Object toValue = valueCastFunction.invoke(fromValue);

--- a/presto-main/src/test/java/com/facebook/presto/type/TestMapOperators.java
+++ b/presto-main/src/test/java/com/facebook/presto/type/TestMapOperators.java
@@ -417,6 +417,15 @@ public class TestMapOperators
         assertFunction("CAST(MAP(ARRAY[1], ARRAY[MAP(ARRAY[1.0], ARRAY[false])]) AS MAP<varchar, MAP(bigint,bigint)>)", new MapType(VARCHAR, new MapType(BIGINT, BIGINT)), ImmutableMap.of("1", ImmutableMap.of(1L, 0L)));
         assertFunction("CAST(MAP(ARRAY[1,2], ARRAY[DATE '2016-01-02', DATE '2016-02-03']) AS MAP(bigint, varchar))", new MapType(BIGINT, VARCHAR), ImmutableMap.of(1L, "2016-01-02", 2L, "2016-02-03"));
         assertFunction("CAST(MAP(ARRAY[1,2], ARRAY[TIMESTAMP '2016-01-02 01:02:03', TIMESTAMP '2016-02-03 03:04:05']) AS MAP(bigint, varchar))", new MapType(BIGINT, VARCHAR), ImmutableMap.of(1L, "2016-01-02 01:02:03.000", 2L, "2016-02-03 03:04:05.000"));
+
+        // null values
+        Map<Long, Double> expected = new HashMap<>();
+        expected.put(0L, 1.0);
+        expected.put(1L, null);
+        expected.put(2L, null);
+        expected.put(3L, 2.0);
+        assertFunction("CAST(MAP(ARRAY[0, 1, 2, 3], ARRAY[1,NULL, NULL, 2]) AS MAP<BIGINT, DOUBLE>)", new MapType(BIGINT, DOUBLE), expected);
+
         assertInvalidCast("CAST(MAP(ARRAY[1, 2], ARRAY[6, 9]) AS MAP<boolean, bigint>)", "duplicate keys");
     }
 }

